src/hb-iter.hh

```diff
@@ -627,10 +627,12 @@ struct hb_concat_iter_t :
   void __forward__ (unsigned n)
   {
     if (n > a.len ()) {
-      a.__forward__ (a.len ());
       n -= a.len ();
+      a.__forward__ (a.len ());
+      b.__forward__ (n);
```

Comment:
I don't like that you are calling `a.len()` three times. It might be `O(n)`.

Reply:
Makes sense. I add an alternate implementation when the iterator is not random access.